### ISSUE TITLE
How can id run multi-node training?

### ISSUE THREAD

**Issue by mandal4** [ultralytics/ultralytics#16892]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi there,
How can i run multi-node training?
As i searched prev-issues, it seems not to be in official. 
I tried it with following prev-issues, but failed it. 
Is there someone success on it?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mandal4, thank you for reaching out to Ultralytics 🚀! 

We recommend checking out the [Docs](https://docs.ultralytics.com) to explore various usage scenarios which might help you with multi-node training.

If this is a custom training ❓ Question, please provide detailed information, including any attempted configurations, error messages, and logs. This will help us and the community assist you better.

For real-time discussions, join our community on [Discord](https://ultralytics.com/discord) 🎧. Alternatively, participate in detailed discussions on [Discourse](https://community.ultralytics.com) or join our [Subreddit](https://reddit.com/r/ultralytics) for more insights and shared experiences.

## Upgrade

Please confirm you are using the latest version of `ultralytics` by running:

```bash
pip install -U ultralytics
```

Ensure your environment aligns with our [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), using [**Python>=3.8**](https://www.python.org/) and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/).

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

A green badge indicates that all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are successful. This verifies the reliable operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across various platforms.

This is an automated response 😊. An Ultralytics engineer will reach out soon to provide further assistance.

**Comment by glenn-jocher** [comment]
@mandal4 hi, multi-node training isn't officially supported in the current Ultralytics YOLO framework. For distributed training, consider using PyTorch's native distributed data parallel (DDP) features.

**Comment by sjtu-cz** [comment]
> [@mandal4](https://github.com/mandal4) hi, multi-node training isn't officially supported in the current Ultralytics YOLO framework. For distributed training, consider using PyTorch's native distributed data parallel (DDP) features.

Can you provide a sample code?

**Comment by sjtu-cz** [comment]
> [@mandal4](https://github.com/mandal4) hi, multi-node training isn't officially supported in the current Ultralytics YOLO framework. For distributed training, consider using PyTorch's native distributed data parallel (DDP) features.

Can you provide a sample code?

**Comment by glenn-jocher** [comment]
@sjtu-cz for multi-node training with PyTorch DDP, you can structure your script like this:

```python
import torch
from ultralytics import YOLO

# Initialize distributed training
torch.distributed.init_process_group(backend='nccl')
local_rank = int(os.environ['LOCAL_RANK'])
torch.cuda.set_device(local_rank)

# Load model and train
model = YOLO('yolo11n.pt').to(local_rank)
model.train(data='coco8.yaml', epochs=100, device=local_rank)
```

Launch it using PyTorch's distributed launcher:
```bash
torchrun --nproc_per_node=N --nnodes=M --node_rank=INDEX --master_addr=ADDR train.py
```

For full implementation details, refer to PyTorch's official Distributed Data Parallel documentation.

**Comment by sjtu-cz** [comment]
您好，你的邮件我已经收到~

**Comment by glenn-jocher** [comment]
For multi-node training with Ultralytics YOLO11, you can implement a PyTorch DDP setup like this minimal example:

```python
import os
import torch
from ultralytics import YOLO

def main():
    torch.distributed.init_process_group(backend='nccl')
    local_rank = int(os.environ['LOCAL_RANK'])
    torch.cuda.set_device(local_rank)
    
    model = YOLO('yolo11n.yaml').to(local_rank)
    model.train(data='coco8.yaml', epochs=100, device=local_rank)

if __name__ == "__main__":
    main()
```

Launch with:
```bash
torchrun --nnodes=2 --node_rank=0 --nproc_per_node=8 --master_addr=<MASTER_IP> --master_port=1234 train.py
```

This requires proper network configuration between nodes. For implementation details, refer to PyTorch's official [Distributed Data Parallel documentation](https://pytorch.org/docs/stable/distributed.html). This is an advanced setup that may require additional cluster-specific configuration.

**Comment by sjtu-cz** [comment]
Can this support multi-node with each node 8-card training?
I used the above code, but each machine only has one graphics card in use. and the batch_size with above 8 will OOM.


> For multi-node training with Ultralytics YOLO11, you can implement a PyTorch DDP setup like this minimal example:
> 
> import os
> import torch
> from ultralytics import YOLO
> 
> def main():
>     torch.distributed.init_process_group(backend='nccl')
>     local_rank = int(os.environ['LOCAL_RANK'])
>     torch.cuda.set_device(local_rank)
>     
>     model = YOLO('yolo11n.yaml').to(local_rank)
>     model.train(data='coco8.yaml', epochs=100, device=local_rank)
> 
> if __name__ == "__main__":
>     main()
> Launch with:
> 
> torchrun --nnodes=2 --node_rank=0 --nproc_per_node=8 --master_addr=<MASTER_IP> --master_port=1234 train.py
> This requires proper network configuration between nodes. For implementation details, refer to PyTorch's official [Distributed Data Parallel documentation](https://pytorch.org/docs/stable/distributed.html). This is an advanced setup that may require additional cluster-specific configuration.

**Comment by sjtu-cz** [comment]
> [@glenn-jocher](https://github.com/glenn-jocher) Can this support multi-node with each node 8-card training? I used the above code, but each machine only has one graphics card in use. and the batch_size with above 8 will OOM.

refer this issue， https://github.com/ultralytics/ultralytics/issues/7038 ，Everything is going smoothly at the moment.

**Comment by glenn-jocher** [comment]
Thank you for the update! For multi-node training with YOLO11, while not officially supported, you can implement it through PyTorch's native DDP. Ensure your launch command includes all nodes and proper networking:

```bash
torchrun --nnodes=2 --nproc_per_node=8 --node_rank=0 --master_addr=<MASTER_IP> --master_port=1234 train.py
```

If encountering OOM, reduce `batch_size` per GPU or use `batch=-1` for [AutoBatch](https://docs.ultralytics.com/guides/hyperparameter-tuning#autobatch) to find the optimal size. For implementation details, refer to PyTorch's official [Distributed Training documentation](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html).

**Comment by sjtu-cz** [comment]
> Thank you for the update! For multi-node training with YOLO11, while not officially supported, you can implement it through PyTorch's native DDP. Ensure your launch command includes all nodes and proper networking:
> 
> torchrun --nnodes=2 --nproc_per_node=8 --node_rank=0 --master_addr=<MASTER_IP> --master_port=1234 train.py
> If encountering OOM, reduce `batch_size` per GPU or use `batch=-1` for [AutoBatch](https://docs.ultralytics.com/guides/hyperparameter-tuning#autobatch) to find the optimal size. For implementation details, refer to PyTorch's official [Distributed Training documentation](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html).

However, I found that the accuracy of multi-machine training cannot align with that of single-machine training. Do I need to adjust the learning rate or make changes elsewhere?
